File: solutions/3622-check-divisibility-by-digit-sum-and-product/math.cpp
class Solution {
  public:
    bool checkDivisibility(int n) {
        int s = 0;
        int p = 1;
        const int num = n;

        while (n) {
            const int digit = n % 10;
            p *= digit;
            s += digit;
            n /= 10;
        }

        return num % (p + s) == 0;
    }
};

File: solutions/3622-check-divisibility-by-digit-sum-and-product/math.js
// time: O(n)
// space: O(1)

/**
 * @param {number} n
 * @return {boolean}
 */
const checkDivisibility = n => {
    const num = n;
    let p = 1;
    let s = 0;

    while (n) {
        const digit = n % 10;

        p *= digit;
        s += digit;
        n = Math.floor(n / 10);
    }

    return num % (s + p) === 0;
};

